### PR TITLE
IDEMPIERE-4907 : Jasper Report using Dashboard Content

### PR DESCRIPTION
**Pull request by uthadehikaru**
https://idempiere.atlassian.net/browse/IDEMPIERE-4907

# Pull Request Checklist

- [x] My code follows the [code guidelines](https://wiki.idempiere.org/en/Contributing_to_Trunk) of this project
- [x] My code follows the best practices of this project
- [x] I have performed a self-review of my own code
- [x] My code is easy to understand and review. 
- [x] I have checked my code and corrected any misspellings
- [ ] In hard-to-understand areas, I have commented my code.
- [x] My changes generate no new warnings
### Tests
- [x] I have tested the direct scenario that my code is solving
- [ ] I checked all collaterals that can be affected by my changes, and tested other potential affected scenarios
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added unit tests that prove my fix is effective or that my feature works
### Documentation
- [ ] I have made corresponding changes to the documentation as follows:
- - [ ] New feature (non-breaking change which adds functionality): I have created the New Feature page in the project wiki explaining the functionality and how to use it. If relevant, I have committed sample data to the core seed to have usable examples in GardenWorld.
- - [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected): I have documented the change in a clear way that everyone in the community can understand the impact of the change.
- - [ ] Improvement (improves and existing functionality): This documentation is needed if the improvement changes the way the user interacts with the system or the outcome of a process/task changes. If it is just, for instance, a performance improvement, documentation might not be needed. 
- [ ] The changed/added documentation is in the project wiki (not privately-hosted pdf files or links pointing to a company website) and is complete and self-explanatory.

### CODE REVIEW

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java [hunk 1]
```diff
@@ -1612,13 +1618,20 @@ private ReportEngine runReport(int AD_Process_ID, int AD_PrintFormat_ID, String
...
 			pi.setAD_User_ID(Env.getAD_User_ID(Env.getCtx()));
 			pi.setAD_Client_ID(Env.getAD_Client_ID(Env.getCtx()));
 			pi.setAD_PInstance_ID(pInstance.getAD_PInstance_ID());		
 			if (!process.processIt(pi, null) && pi.getClassName() != null) 
 				throw new IllegalStateException("Process failed: (" + pi.getClassName() + ") " + pi.getSummary());
 		
 			//	Report
+			if(process.getJasperReport() != null) {
+				ServerProcessCtl.process(pi, null);
+			}
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:1634]: this part seems wrong, you either have JasperReport or ReportEngine, you can't have both here.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java [hunk 2]
```diff
@@ -1598,7 +1604,7 @@ private String stripHtml(String htmlString, boolean all) {
 	 */
 	private ReportEngine runReport(int AD_Process_ID, int AD_PrintFormat_ID, String parameters) {
    		MProcess process = MProcess.get(Env.getCtx(), AD_Process_ID);
-		if (!process.isReport() || process.getAD_ReportView_ID() == 0)
+		if (!process.isReport())
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:1607]: why the need to remove the checking here ? Isn't JasperReport never reach here ?

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java [hunk 3]
```diff
@@ -1643,13 +1656,51 @@ private ReportEngine runReport(int AD_Process_ID, int AD_PrintFormat_ID, String
...
+	private File runJasperReport(MProcess process, String parameters) {
+		MPInstance pInstance = new MPInstance(Env.getCtx(), process.getAD_Process_ID(), 0, 0, null);
+		pInstance.setIsProcessing(true);
+		pInstance.saveEx();
+		try {
+			if(!fillParameter(pInstance, parameters))
+				return null;
+			//
+			ProcessInfo pi = new ProcessInfo (process.getName(), process.getAD_Process_ID(), 0, 0);
+			if(process.getJasperReport() != null) {
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:1684]: getJasperReport can't be null at this point. should throw runtime exception if it is null.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java [hunk 4]
```diff
@@ -1643,13 +1656,51 @@ private ReportEngine runReport(int AD_Process_ID, int AD_PrintFormat_ID, String
...
+				pi.setExportFileExtension("html");
+			}
+			pi.setAD_User_ID(Env.getAD_User_ID(Env.getCtx()));
+			pi.setAD_Client_ID(Env.getAD_Client_ID(Env.getCtx()));
+			pi.setAD_PInstance_ID(pInstance.getAD_PInstance_ID());		
+			if (!process.processIt(pi, null) && pi.getClassName() != null) 
+				throw new IllegalStateException("Process failed: (" + pi.getClassName() + ") " + pi.getSummary());
+		
+			//	Report
+			if(process.getJasperReport() != null) {
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:1695]: same as above, getJasperReport can't be null here.

**Review by hengsin — CHANGES_REQUESTED** [review]

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java (hunk 1 above)
Comment (uthadehikaru) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:1634]: i see. it's my first approach when trying to run jasper report and then i think i need to create separated method so i create runJasperReport method

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java (hunk 2 above)
Comment (uthadehikaru) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:1607]: yup, this my first approach also

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java (hunk 3 above)
Comment (uthadehikaru) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:1684]: so the logic is if getJasperReport is null, then throw exception, instead run ServerProcessCtl if jasperReport is not null. got it

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java [hunk 5]
```diff
@@ -1643,13 +1649,48 @@ private ReportEngine runReport(int AD_Process_ID, int AD_PrintFormat_ID, String
...
+	private File runJasperReport(MProcess process, String parameters) {
+		MPInstance pInstance = new MPInstance(Env.getCtx(), process.getAD_Process_ID(), 0, 0, null);
+		pInstance.setIsProcessing(true);
+		pInstance.saveEx();
+		try {
+			if(!fillParameter(pInstance, parameters))
+				return null;
+			//
+			if(process.getJasperReport() == null)
+				throw new RuntimeException("no Jasper Report File provided. Process "+process.getName());
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:1677]: This type of checking should be first in the method (i.e before the MPInstance pInstance line). However, looking at the flow of the code, isn't this should never happens (you already has the if(process.getJasperReport() != null) checking above).

**Review by hengsin — CHANGES_REQUESTED** [review]

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java (hunk 5 above)
Comment (uthadehikaru) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:1677]: actually, on line 1654, i already check it. so if getJasperReport is null, the the method runJasperReport not execute. should i still add check getJasperReport on first line of runJasperReport method or just remove check getJasperReport on code above?

Comment (hengsin) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/desktop/DashboardController.java:1677]: yes, I means the checking here is redundant.

**Review by hengsin — APPROVED** [review]
looks good

**Pull request merged**